server/main_server/src/main.cpp

```diff
@@ -1450,6 +1567,133 @@ Environment Variables:
+            input.dataObjInfo = &info;
+            input.regParam = &kvp;
+
+            for (int i = 0; i < 250; ++i) {
+                if (!irods::access_time_manager::try_dequeue(data)) {
+                    break;
+                }
+
+                log_server::debug("{}: Applying access time update: data_id=[{}], replica_number=[{}], last_accessed=[{}]",
+                    __func__, data.data_id, data.replica_number, data.last_accessed);
+
+                info.dataId = data.data_id;
+                info.replNum = data.replica_number;
+
+                irods::at_scope_exit free_kvp{[&kvp] { clearKeyVal(&kvp); }};
+                const auto new_atime = fmt::format("{:011}", data.last_accessed);
+                addKeyVal(&kvp, DATA_ACCESS_TIME_KW, new_atime.c_str());
+                addKeyVal(&kvp, ADMIN_KW, "");
+
+                const auto ec = rcModDataObjMeta(static_cast<RcComm*>(conn), &input);
+                log_server::debug("{}: rcModDataObjMeta returned [{}]", __func__, ec);
+            }
+#else
+            irods::experimental::client_connection conn;
+            irods::access_time_manager::access_time_data data;
+            nlohmann::json::array_t updates;
+
+            const auto update_count = irods::access_time_manager::number_of_queued_updates();
+            for (std::size_t i = 0; i < update_count; ++i) {
+                if (!irods::access_time_manager::try_dequeue(data)) {
+                    break;
+                }
+
+                updates.push_back(nlohmann::json{
+                    {"data_id", data.data_id},
+                    {"replica_number", data.replica_number},
+                    {"atime", data.last_accessed}
+                });
+            }
+
+            if (updates.empty()) {
+                return;
+            }
+
+            // Group elements by data id and replica number in descending order.
+            std::sort(std::begin(updates), std::end(updates), [](const auto& _lhs, const auto& _rhs) {
+                const auto& lhs_data_id = _lhs.at("data_id");
+                const auto& rhs_data_id = _rhs.at("data_id");
+                if (lhs_data_id > rhs_data_id) {
+                    return true;
+                }
+
+                if (lhs_data_id == rhs_data_id) {
+                    const auto& lhs_replica_number = _lhs.at("replica_number");
+                    const auto& rhs_replica_number = _rhs.at("replica_number");
+                    if (lhs_replica_number > rhs_replica_number) {
+                        return true;
+                    }
+
+                    if (lhs_replica_number == rhs_replica_number) {
+                        return _lhs.at("atime") > _rhs.at("atime");
+                    }
+                }
+
+                return false;
+            });
+
+            // Reduce updates such that each replica is updated once. On completion, the JSON
+            // is guaranteed to contain only the most recent update for each replica.
```

Comment:
could have a debug metric here too, how big before and after reduction.

Reply:
See https://github.com/irods/irods/pull/8269#issuecomment-2725796466.